store/sqlstore/integrity.go

```diff
@@ -0,0 +1,504 @@
+// Copyright (c) 2019-present Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+package sqlstore
+
+import (
+	"fmt"
+
+	"github.com/mattermost/gorp"
+	"github.com/mattermost/mattermost-server/mlog"
+	"github.com/mattermost/mattermost-server/store"
+)
+
+type relationalCheckConfig struct {
+	parentName         string
+	parentIdAttr       string
+	childName          string
+	childIdAttr        string
+	canParentIdBeEmpty bool
+	sortRecords        bool
+}
+
+func getOrphanedRecords(dbmap *gorp.DbMap, cfg relationalCheckConfig) ([]store.OrphanedRecord, error) {
+	var records []store.OrphanedRecord
+
+	query := fmt.Sprintf(`SELECT %s AS ParentId`, cfg.parentIdAttr)
+
+	if cfg.childIdAttr != "" {
+		query += fmt.Sprintf(` , %s AS ChildId`, cfg.childIdAttr)
+	}
+
+	query += fmt.Sprintf(`
+		FROM
+			%s
+		WHERE NOT EXISTS (
+			SELECT
+  			id
+			FROM
+				%s
+			WHERE
+				id = %s.%s
+		)
+	`, cfg.childName, cfg.parentName, cfg.childName, cfg.parentIdAttr)
+
+	if cfg.canParentIdBeEmpty {
+		query += fmt.Sprintf(` AND %s != ''`, cfg.parentIdAttr)
+	}
+
+	if cfg.sortRecords {
+		query += fmt.Sprintf(` ORDER BY %s`, cfg.parentIdAttr)
+	}
+
+	_, err := dbmap.Select(&records, query)
+
+	return records, err
+}
+
+func checkParentChildIntegrity(dbmap *gorp.DbMap, config relationalCheckConfig) store.IntegrityCheckResult {
+	var result store.IntegrityCheckResult
+	var data store.RelationalIntegrityCheckData
+
+	config.sortRecords = true
+	data.Records, result.Err = getOrphanedRecords(dbmap, config)
+	if result.Err != nil {
+		mlog.Error(result.Err.Error())
+		return result
+	}
+	data.ParentName = config.parentName
+	data.ChildName = config.childName
+	data.ParentIdAttr = config.parentIdAttr
+	data.ChildIdAttr = config.childIdAttr
+	result.Data = data
+
+	return result
+}
+
+func checkChannelsCommandWebhooksIntegrity(dbmap *gorp.DbMap) store.IntegrityCheckResult {
+	var config relationalCheckConfig
```

Comment:
Instead of assigning (which duplicates `config.` frequently), let's format as follows. Note that I'd also recommend listing the child first and the parent second, since I think it's more natural to imagine the relationship in that direction.
```go
return checkParentChildIntegrity(dbmap, relationalCheckConfig{
	childName: "CommandWebhooks",
	childIdAttr: "Id",
    parentName: "Channels",
    parentIdAttr: "ChannelId",
})
```

Reply:
Yes, I tend to be quite explicit with the downside of repeating a bit of code. I will update.

As far as the relationship it was an explicit choice to keep it from parent to child and not vice versa.
My line of thought was that the parent relationships are the ones that when missing create the consistency problem we check for and hence grouping by parent relationship made more sense. I understand it can get confusing though.